pajbot/web/routes/base/login.py

```diff
@@ -166,3 +195,67 @@ def logout():
     @twitch.tokengetter
     def get_twitch_oauth_token():
         return session.get("twitch_token")
+
+    if not spotify:
+        return
+
+    @app.route("/spotify_login")
+    def spotify_login():
+        if spotify is not None:
+            callback_url = (
+                app.bot_config["spotify"]["redirect_uri"]
+                if "redirect_uri" in app.bot_config["spotify"]
+                else url_for("authorized", _external=True)
+            )
+            state = request.args.get("n") or request.referrer or None
+            return spotify.authorize(
+                callback=callback_url, state=state, scope=" ".join(spotify_scopes), force_verify="true"
+            )
+        return render_template("login_error.html")
+
+    @app.route("/login/spotify_auth")
+    def spotify_auth():
+        try:
+            resp = spotify.authorized_response(spotify=True)
+        except OAuthException as e:
+            log.error(e)
+            log.exception("An exception was caught while authorizing")
+            next_url = get_next_url(request, "state")
+            return redirect(next_url)
+        except Exception as e:
+            log.error(e)
+            log.exception("Unhandled exception while authorizing")
+            return render_template("login_error.html")
+
+        session["spotify_token"] = (resp["access_token"],)
+        if resp is None:
+            if "error" in request.args and "error_description" in request.args:
+                log.warning(f"Access denied: reason={request.args['error']}, error={request.args['error_description']}")
+            next_url = get_next_url(request, "state")
+            return redirect(next_url)
+        elif type(resp) is OAuthException:
+            log.warning(resp.message)
+            log.warning(resp.data)
+            log.warning(resp.type)
+            next_url = get_next_url(request, "state")
+            return redirect(next_url)
+
+        data = f'{app.bot_config["spotify"]["client_id"]}:{app.bot_config["spotify"]["client_secret"]}'
+        encoded = str(base64.b64encode(data.encode("utf-8")), "utf-8")
+        headers = {"Authorization": f"Basic {encoded}"}
+
+        me_api_response = spotify.get("me", headers=headers)
+
+        redis = RedisManager.get()
+        token_json = UserAccessToken.from_api_response(resp).jsonify()
+
+        redis.set(f"authentication:spotify-access-token:{me_api_response.data['id']}", json.dumps(token_json))
+        redis.set(f"authentication:user-refresh-token:{me_api_response.data['id']}", token_json["refresh_token"])
```

Comment:
the refresh token should be included with the JSON under `spotify-access-token`, just like how the twitch tokens are stored

also `user-refresh-token`, if it was to stay, should also be namespaced to signify it belongs to spotify